extensions/indexes/lucene/src/main/java/org/exist/xquery/modules/lucene/Field.java

```diff
@@ -84,13 +115,134 @@ public Sequence eval(Sequence[] args, Sequence contextSequence) throws XPathExce
+        final Map<Object, Query> terms = index.getTerms(match.getQuery());
+        final NodePath path = LuceneMatchListener.getPath(proxy);
+        final LuceneConfig config = index.getLuceneConfig(context.getBroker(), proxy.getDocumentSet());
+        LuceneIndexConfig idxConf = config.getConfig(path).next();
+        if (idxConf == null) {
+            // no lucene index: no fields to highlight
+            return Sequence.EMPTY_SEQUENCE;
+        }
+
+        final Analyzer analyzer = idxConf.getAnalyzer();
+
+        context.pushDocumentContext();
+        try {
+            final MemTreeBuilder builder = context.getDocumentBuilder();
+            builder.startDocument();
+
+            final InMemoryNodeSet result =  new InMemoryNodeSet(text.getItemCount());
+            for (final SequenceIterator si = text.iterate(); si.hasNext(); ) {
+                final int nodeNr = builder.startElement(Namespaces.EXIST_NS, "field", "exist:field", null);
+                final String content = si.nextItem().getStringValue();
+                int currentPos = 0;
+                try (final Reader reader = new StringReader(content);
+                     final TokenStream tokenStream = analyzer.tokenStream(fieldName, reader)) {
+                    tokenStream.reset();
+                    final MarkableTokenFilter stream = new MarkableTokenFilter(tokenStream);
+                    while (stream.incrementToken()) {
+                        String token = stream.getAttribute(CharTermAttribute.class).toString();
+                        final Query query = terms.get(token);
+                        if (query != null) {
+                            if (match.getQuery() instanceof PhraseQuery) {
+                                final Term phraseTerms[] = ((PhraseQuery) match.getQuery()).getTerms();
+                                if (token.equals(phraseTerms[0].text())) {
+                                    // Scan the following text and collect tokens to see
+                                    // if they are part of the phrase.
+                                    stream.mark();
+                                    int t = 1;
+                                    OffsetAttribute offset = stream.getAttribute(OffsetAttribute.class);
+                                    final int startOffset = offset.startOffset();
+                                    int endOffset = offset.endOffset();
+                                    while (stream.incrementToken() && t < phraseTerms.length) {
+                                        token = stream.getAttribute(CharTermAttribute.class).toString();
+                                        if (token.equals(phraseTerms[t].text())) {
+                                            offset = stream.getAttribute(OffsetAttribute.class);
+                                            endOffset = offset.endOffset();
+                                            if (++t == phraseTerms.length) {
```

Comment:
++t on purpose? for readability it is better to have it as a separate expression no?

Reply:
I agree, it is definitely more readable to have it in a separate expression. The subtle difference between `++t` and `t++` is easy to miss.

Reply:
for me it is.